src/modules/netinstall/PackageModel.cpp

```diff
@@ -0,0 +1,266 @@
+        itemMap.insert( "pre-script", item->parentItem()->preScript() ); // Only groups have hooks
+        itemMap.insert( "package", item->packageName() );
+        itemMap.insert( "post-script", item->parentItem()->postScript() );
+        packages.append( QVariant( itemMap ) );
+    }
+    return packages;
+}
+
+QList<PackageTreeItem*>
+PackageModel::getItemPackages( PackageTreeItem* item, bool isCritical ) const
+{
+    QList<PackageTreeItem*> selectedPackages;
+    for ( int i = 0; i < item->childCount(); i++ )
+    {
+        if ( item->child( i )->isSelected() == Qt::Unchecked  || 
+                item->child( i )->isCritical() != isCritical )
+            continue;
+
+        if ( !item->child( i )->childCount() ) // package
+            selectedPackages.append( item->child( i ) );
+        else
+            selectedPackages.append( getItemPackages( item->child( i ), isCritical ) );
+    }
+    return selectedPackages;
+
+}
+
+void
+PackageModel::setupModelData( const YAML::Node& data, PackageTreeItem* parent )
+{
+    for ( YAML::const_iterator it = data.begin(); it != data.end(); ++it )
+    {
+        const YAML::Node itemDefinition = *it;
+
+        QString name(
+                tr( CalamaresUtils::yamlToVariant( itemDefinition["name"] ).toByteArray() ) );
+        QString description(
+                tr( CalamaresUtils::yamlToVariant( itemDefinition["description"] ).toByteArray() ) );
+
+        PackageTreeItem::ItemData itemData;
+        itemData.name = name;
+        itemData.description = description;
+
+        if ( itemDefinition["pre-install"] )
+            itemData.preScript =
+                CalamaresUtils::yamlToVariant( itemDefinition["pre-install"] ).toString();
+        if ( itemDefinition["post-install"] )
+            itemData.postScript =
+                CalamaresUtils::yamlToVariant( itemDefinition["post-install"] ).toString();
+        PackageTreeItem* item = new PackageTreeItem(itemData, parent );
+
+        if ( itemDefinition["selected"] )
+            item->setSelected(
+                    CalamaresUtils::yamlToVariant( itemDefinition["selected"] ).toBool() ?
+                    Qt::Checked : Qt::Unchecked );
+        else
+            item->setSelected( parent->isSelected() ); // Inherit from it's parent
+
+        if ( itemDefinition["hidden"] )
+            item->setHidden(
+                CalamaresUtils::yamlToVariant( itemDefinition["hidden"] ).toBool() );
+
+        if ( itemDefinition["critical"] )
+            item->setCritical(
+                CalamaresUtils::yamlToVariant( itemDefinition["critical"] ).toBool() );
+
+        for ( YAML::const_iterator packageIt = itemDefinition["packages"].begin();
+                packageIt != itemDefinition["packages"].end(); ++packageIt )
+            item->appendChild( 
+                    new PackageTreeItem( CalamaresUtils::yamlToVariant( *packageIt ).toString(), item ) );
+
+        if ( itemDefinition["subgroups"] )
```

Comment:
Ugliness is in the eye of the beholder, and that's something the maintainer of the distribution's netinstall group list YAML file will have to sort out. This is not something the end user can mess with. So I don't see a point in hardcoding a limit here.